File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/converter/TestDataFromIfStatementConverter.java
package org.skellig.teststep.processing.converter;

import org.apache.commons.lang3.math.NumberUtils;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import java.util.Map;
import java.util.Objects;

class TestDataFromIfStatementConverter implements TestDataConverter {

    private static final String IF_KEYWORD = "if";
    private static final String CONDITION_KEYWORD = "condition";
    private static final String THEN_KEYWORD = "then";
    private static final String ELSE_KEYWORD = "else";

    private ScriptEngine engine = new ScriptEngineManager().getEngineByName("JavaScript");

    public Object convert(Object value) {
        if (value instanceof Map) {
            Map<String, Object> valueAsMap = (Map<String, Object>) value;
            if (valueAsMap.containsKey(IF_KEYWORD)) {
                Map<String, Object> ifDetails = (Map<String, Object>) valueAsMap.get(IF_KEYWORD);
                String condition = (String) ifDetails.get(CONDITION_KEYWORD);
                Object thenContent = ifDetails.get(THEN_KEYWORD);
                Object elseContent = ifDetails.getOrDefault(ELSE_KEYWORD, "");

                Objects.requireNonNull(condition, "'condition' is mandatory in 'if' statement");
                Objects.requireNonNull(thenContent, "'then' is mandatory in 'if' statement");

                return isConditionSatisfied(condition) ? thenContent : elseContent;
            }
        }
        return value;
    }

    private boolean isConditionSatisfied(String condition) {
        StringBuilder newCondition = encloseStringWithQuotes(condition);

        try {
            return (Boolean) engine.eval(newCondition.toString());
        } catch (Exception e) {
            return false;
        }
    }

    private StringBuilder encloseStringWithQuotes(String condition) {
        StringBuilder newCondition = new StringBuilder();

        for (int i = 0; i < condition.length(); i++) {
            char c = condition.charAt(i);
            if (isSpecialSymbol(c)) {
                newCondition.append(c);
            } else {
                String value = "";
                for (; i < condition.length(); i++) {
                    c = condition.charAt(i);
                    if (!isSpecialSymbol(c)) {
                        value += c;
                    } else {
                        i--;
                        break;
                    }
                }
                value = value.trim();
                newCondition.append(value.equals("") || NumberUtils.isNumber(value) ? value : "'" + value + "'");
            }
        }
        return newCondition;
    }

    private boolean isSpecialSymbol(char c) {
        return c == '(' || c == ')' || c == '<' || c == '=' || c == '>' || c == '&' || c == '|';
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/converter/TestDataFromIfStatementConverter.kt
package org.skellig.teststep.processing.converter

import org.apache.commons.lang3.math.NumberUtils
import java.util.*
import javax.script.ScriptEngineManager

class TestDataFromIfStatementConverter : TestStepValueConverter {

    companion object {
        private const val IF_KEYWORD = "if"
        private const val CONDITION_KEYWORD = "condition"
        private const val THEN_KEYWORD = "then"
        private const val ELSE_KEYWORD = "else"
    }

    private val engine = ScriptEngineManager().getEngineByName("JavaScript")

    override fun convert(value: Any?): Any? =
            when (value) {
                is Map<*, *> ->
                    if (value.containsKey(IF_KEYWORD)) {
                        val ifDetails = value[IF_KEYWORD] as Map<String, Any>
                        val condition = ifDetails[CONDITION_KEYWORD] as String?
                        val thenContent = ifDetails[THEN_KEYWORD]
                        val elseContent = ifDetails[ELSE_KEYWORD] ?: ""

                        Objects.requireNonNull(condition, "'condition' is mandatory in 'if' statement")
                        Objects.requireNonNull(thenContent, "'then' is mandatory in 'if' statement")

                        if (isConditionSatisfied(condition)) convert(thenContent)
                        else convert(elseContent)
                    } else value.entries.map { it.key to convert(it.value) }.toMap()

                is Collection<*> -> value.map { convert(it) }.toList()
                else -> value
            }

    private fun isConditionSatisfied(condition: String?): Boolean {
        val newCondition = encloseStringWithQuotes(condition)
        return try {
            engine.eval(newCondition.toString()) as Boolean
        } catch (e: Exception) {
            false
        }
    }

    private fun encloseStringWithQuotes(condition: String?): StringBuilder {
        val newCondition = StringBuilder()
        var i = 0
        while (i < condition!!.length) {
            var c = condition[i]
            if (isSpecialSymbol(c)) {
                newCondition.append(c)
            } else {
                var value = ""
                while (i < condition.length) {
                    c = condition[i]
                    value += if (!isSpecialSymbol(c)) {
                        c
                    } else {
                        i--
                        break
                    }
                    i++
                }
                value = value.trim { it <= ' ' }
                newCondition.append(if (value == "" || NumberUtils.isNumber(value)) value else "'$value'")
            }
            i++
        }
        return newCondition
    }

    private fun isSpecialSymbol(c: Char): Boolean {
        return c == '(' || c == ')' || c == '<' || c == '=' || c == '>' || c == '&' || c == '|'
    }
}